ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/transactions/layered/LayeredPendingTransactions.java

```diff
@@ -321,64 +321,69 @@ public synchronized void selectTransactions(
+                            !alreadyChecked.contains(candidatePendingTx.getHash())
+                                && candidatePendingTx.getNonce() <= highPrioPendingTx.getNonce())
                     .forEach(
                         candidatePendingTx -> {
                           alreadyChecked.add(candidatePendingTx.getHash());
-                          switch (selector.evaluateTransaction(
-                              candidatePendingTx.getTransaction())) {
-                            case CONTINUE:
-                              LOG.atTrace()
-                                  .setMessage("CONTINUE: Transaction {}")
-                                  .addArgument(candidatePendingTx::toTraceLog)
-                                  .log();
-                              break;
-                            case DELETE_TRANSACTION_AND_CONTINUE:
-                              invalidTransactions.add(candidatePendingTx);
-                              LOG.atTrace()
-                                  .setMessage("DELETE_TRANSACTION_AND_CONTINUE: Transaction {}")
-                                  .addArgument(candidatePendingTx::toTraceLog)
-                                  .log();
-                              logTransactionForReplayDelete(candidatePendingTx);
-                              break;
-                            case COMPLETE_OPERATION:
-                              completed.set(true);
-                              LOG.atTrace()
-                                  .setMessage("COMPLETE_OPERATION: Transaction {}")
-                                  .addArgument(candidatePendingTx::toTraceLog)
-                                  .log();
-                              break;
+                          final var res =
+                              selector.evaluateTransaction(candidatePendingTx.getTransaction());
+
+                          LOG.atTrace()
+                              .setMessage("Selection result {} for transaction {}")
+                              .addArgument(res)
+                              .addArgument(candidatePendingTx::toTraceLog)
+                              .log();
+
+                          if (res.discard()) {
+                            invalidTransactions.add(candidatePendingTx);
+                            logTransactionForReplayDelete(candidatePendingTx);
+                          }
+
+                          if (res.stop()) {
+                            completed.set(true);
+                          }
+
+                          if (!res.selected()) {
+                            // avoid processing other txs from this sender if this one is skipped
+                            // since the following will not be selected due to the nonce gap
+                            skipSenders.add(candidatePendingTx.getSender());
+                            LOG.trace("Skip sender {}", candidatePendingTx.getSender());
```

Comment:
```suggestion
                            LOG.trace("Skipping tx from sender {}", candidatePendingTx.getSender());
```

Reply:
how does the sender get removed from the skipSenders list?

Reply:
He should not be removed for the remaining of this selection task, since evaluating his txs is wasted time after the first non selected. This list only last for a single block creation, so he could be picked again in a following selection task.